File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
import collections

dx = [1, 2, 2, 1, -1, -2, -2, -1]
dy = [2, 1, -1, -2, -2, -1, 1, 2]

def BFS(startDot):
    q = collections.deque([startDot])
    while q:
        y, x = q.popleft()
        if((y, x) == targetDot):
            break
        for s in range(8):
            newX = x + dx[s]
            newY = y + dy[s]
            if(newX >= 0 and newX < I and newY >= 0 and newY < I):
                if(arr[newY][newX] == 0):
                    arr[newY][newX] = arr[y][x] + 1
                    q.append((newY, newX))

T = int(sys.stdin.readline())

for i in range(T):
    I = int(sys.stdin.readline())
    arr = [[0 for i in range(I)] for i in range(I)]
    
    # 앞이 y축 뒤가 x축이라 생각
    currentDot = tuple(map(int, sys.stdin.readline().split(" ")))
    targetDot = tuple(map(int, sys.stdin.readline().split(" ")))
    BFS(currentDot)
    print(arr[targetDot[0]][targetDot[1]])

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.util.*;
import java.io.*;


public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static int[] dx = {1, 2, 2, 1, -1, -2, -2, -1};
    private static int[] dy = {2, 1, -1, -2, -2, -1, 1, 2};

    private static void BFS(int[] start, int[] end, int len) throws IOException {
        int[][] arr = new int[len][len];
        Deque<int[]> q = new LinkedList<>();
        q.add(new int[] {start[0], start[1]});
        while(!q.isEmpty()) {
            int[] dot = q.poll();
            if(dot[0] == end[0] && dot[1] == end[1]) {
                break;
            }
            for(int i = 0 ; i < 8 ; i++) {
                int x = dot[0] + dx[i];
                int y = dot[1] + dy[i];
                if(x >= 0 && x < len && y >= 0 && y < len) {
                    if(arr[y][x] == 0) {
                        arr[y][x] = arr[dot[1]][dot[0]] + 1;
                        q.add(new int[] {x, y});
                    }
                }
            }
        }
        bw.write(Integer.toString(arr[end[1]][end[0]]));
        bw.write("\n");
        

    }

    public static void main(String[] args) throws IOException {
        
        int T = Integer.parseInt(br.readLine());

        // dot배열의 첫번째값을 x좌표, 두번째값을 y좌표로 지정하자 
        for(int i = 0 ; i < T ; i++) {
            int I = Integer.parseInt(br.readLine());
            st = new StringTokenizer(br.readLine(), " ");
            int[] startDot = new int[2];
            for(int j = 0 ; j < startDot.length ; j++) {
                startDot[j] = Integer.parseInt(st.nextToken());
            }
            st = new StringTokenizer(br.readLine(), " ");
            int[] endDot = new int[2];
            for(int j = 0 ; j < endDot.length ; j++) {
                endDot[j] = Integer.parseInt(st.nextToken());
            }

            BFS(startDot, endDot, I);
        }

        bw.flush();

        br.close();
        bw.close();
    }
}